### DISCUSSION TITLE
Support waiting for a batch to complete before shutting down

### DISCUSSION THREAD

**Discussion by JavaScriptBach** [bbc/sqs-consumer#392] [Feature Request]
### Background

Currently `stop()` prevents any future polls and returns immediately. The user is responsible for waiting and handling any pre-existing requests.

It would be great if the library could provide first-class support for this. A common use case is to shut down cleanly as follows:

1. Stop polling for any future messages
2. Wait a certain amount of time for active requests to complete
3. Continue with shutdown

Looking at the code, it seems like I can sort-of accomplish this by listening on "message_received", although it would be complicated to handle batches correctly. In general this seems like code that is best implemented correctly once in the library so customers can get it for free.

### Objectives

Support cleanly shutting down a consumer without having to cancel in-flight requests.

**Comment by nicholasgriffintn** [comment]
Hmmm, I'm not really sure what you mean here.

The stop event is emitted immediately, but everything is actually stopped immediately unless you call it with abort, like so: `consumer.stop({abort:true})`, that would abort all ongoing requests.

If you don't do that than anything ongoing will still happen after the event.

Are you asking for the event to be emitted when these ongoing events have actually stopped?

I'm not actually sure how possible that is as the consumer doesn't really have any idea of the requests that are ongoing, it doesn't track them and it doesn't feel too useful to me honestly.
